File: src/main/java/dragonbook/inter/expr/Unary.java
package dragonbook.inter.expr;

import dragonbook.lexer.Token;
import dragonbook.symbols.Type;

public class Unary extends Op {
    public Expr expr;

    public Unary(Token tok, Expr x) { // handles minus, for ! see Not
        super(tok, null);
        expr = x;
        type = Type.max(Type.INT, expr.type);
        if (type == null) error("type error");
    }

    public String toString() {
        return op + " " + expr;
    }
}

File: src/main/kotlin/dragonbook/inter/expr/Unary.kt
package dragonbook.inter.expr

import dragonbook.error.ParseError
import dragonbook.lexer.Token
import dragonbook.symbols.Type

data class Unary(override val op: Token, val expr: Expr) : Op { // handles minus, for ! see Not

    init {
        validate()
    }

    private fun validate() {
        if (!expr.type.isNumeric()) throw ParseError(node = this)
    }

    override val type: Type
        get() = if (expr.type == Type.FLOAT) Type.FLOAT else Type.INT

    override fun toString(): String = "$op $expr"
}